src/factory.js

```diff
@@ -87,18 +74,39 @@ class Factory {
   }
 
   async _spawnRemote (options) {
-    const res = await ky.post(
-      `${options.endpoint}/spawn`,
-      {
-        json: {
-          ...options,
-          // avoid recursive spawning
-          remote: false,
-          // do not send code refs over http
-          ipfsModule: { ...options.ipfsModule, ref: undefined },
-          ipfsHttpModule: { ...options.ipfsHttpModule, ref: undefined }
+    const opts = {
+      json: {
+        ...options,
+        // avoid recursive spawning
+        remote: false
+      }
+    }
+
+    if (options.ipfsModule) {
+      delete opts.ipfsModule
+
+      if (options.ipfsModule.path) {
+        opts.ipfsModule = {
+          path: options.ipfsModule.path
+          // n.b. no ref property - do not send code refs over http
+        }
+      }
+    }
+
+    if (options.ipfsHttpModule) {
+      delete opts.ipfsHttpModule
+
+      if (options.ipfsHttpModule.path) {
+        opts.ipfsHttpModule = {
+          path: options.ipfsHttpModule.path
+          // n.b. no ref property - do not send code refs over http
         }
       }
+    }
```

Comment:
why do we need all this haven't we already done this below? we just need to remove refs like in the previous version of this block.

Plus its deleting and setting opts.ipfsModule where it should be opts.json.ipfsModule and same for opts.ipfsHttpModule, so if this code made some use case work it wasn't because of this logic.

Reply:
It was resulting in `ipfsModule: {}` being sent over the http api which was then getting fleshed out to `ipfsModule: { path: require.resolve('ipfs'), ref: require('ipfs') }` which breaks under `js-ipfs`

Reply:
these if clauses are very hard to track and error prone because of all the combinations available in ctl

Reply:
Fair point about the deletion of the wrong property though.

Reply:
> these if clauses are very hard to track and error prone because of all the combinations available in ctl

You are quite right.  Here, look - I've removed them: https://github.com/ipfs/js-ipfsd-ctl/pull/454